src/sentry/services/hybrid_cloud/__init__.py

```diff
@@ -0,0 +1,134 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import Dict, Generic, Mapping, Optional, Type, TypeVar, cast
+
+from django.db.models import F
+
+from sentry.silo import SiloMode
+
+
+class ProjectKeyRole(Enum):
+    store = "store"
+    api = "api"
+
+    def as_orm_role(self):
+        from sentry.models import ProjectKey
+
+        if self == ProjectKeyRole.store:
+            return ProjectKey.roles.store
+        else:
+            return ProjectKey.roles.api
+
+
+@dataclass
+class ApiProjectKey:
+    dsn_public: str = ""
+
+
+class InterfaceWithLifecycle(ABC):
+    @abstractmethod
+    def close(self):
+        pass
+
+
+ServiceInterface = TypeVar("ServiceInterface", bound=InterfaceWithLifecycle)
+
+
+class DelegatedBySiloMode(Generic[ServiceInterface]):
+    """
+    Lazily instantiates, not thread safe in this design.
+    """
+
+    _constructors: Mapping[SiloMode, Type[ServiceInterface]]
+    _singletons: Dict[SiloMode, ServiceInterface]
+
+    def __init__(self, mapping: Mapping[SiloMode, Type[ServiceInterface]]):
+        self._constructors = mapping
+        self._singletons = {}
+
+    def __getattr__(self, item: str):
```

Comment:
Similar to the LazyServiceWrapper, but... way simpler, and it works.  The LazyServiceWrapper does not actually allow dynamically swapping out the backend without a whole bunch of extra selector abstractions.  It seemed like, to make `LazyServiceWrapper` work for our use case, a bunch of refactoring would be needed, which I didn't see as useful.  If you have a better approach that's fine, this was again, my fast pass on the goal.

Reply:
The other thing I'm trying to solve in this wrapper implementation:

1.  Ability to sensibly 'replace' the current implementation in a test.  MOST tests should have a default stub implementation, but certain deeper tests should be able to opt in to a real implementation (integration oriented stuff).  This isn't used yet, but it was a spike while trying to fix different problem.
2.  LIFECYCLES.  It's really important to prevent tests from hanging and also to allow cleanup between test runs that we provide sensible lifecycle management (close hook) around every destruction of a service.  this matters a lot because I want to put the region to kafka producer into a service like this, but it needs an explicit close call (works differently than the old kafka producer).  See the conftest changeset in this PR.